### PR TITLE
fix(nextjs): Next 11.1.0 compatibility

### LINKED ISSUE #6576: Build failed from next version 11.0.2-canary.4
## Current Behavior
Build failed from next version 11.0.2-canary.4

## Expected Behavior
Successfully built from next version 11.0.2-canary.4

## Steps to Reproduce
Upgrade next to `11.0.2-canary.4` and build
Maybe this is the cause
https://github.com/vercel/next.js/pull/26734/files#diff-67f482a12dcb711fa80701859c0c541f51ab7cc9968399325c9736929b58d389

### Failure Logs
```
> nx run myapp:build:production 
Cannot find module 'next/dist/next-server/lib/constants'
Require stack:
- /Users/toan/Desktop/frontend/main-frontend/node_modules/@nrwl/next/src/executors/build/build.impl.js
- /Users/toan/Desktop/frontend/main-frontend/node_modules/@nrwl/tao/src/shared/workspace.js
- /Users/toan/Desktop/frontend/main-frontend/node_modules/@nrwl/tao/src/commands/run.js
- /Users/toan/Desktop/frontend/main-frontend/node_modules/@nrwl/tao/index.js
- /Users/toan/Desktop/frontend/main-frontend/node_modules/@nrwl/cli/lib/run-cli.js

———————————————————————————————————————————————

>  NX   ERROR  Running target "myapp:build" failed
```

### Environment
```
  Node : 14.17.3
  OS   : darwin x64
  yarn : 1.22.11
  
  nx : Not Found
  @nrwl/angular : Not Found
  @nrwl/cli : 12.6.3
  @nrwl/cypress : 12.6.3
  @nrwl/devkit : 12.6.3
  @nrwl/eslint-plugin-nx : 12.6.3
  @nrwl/express : Not Found
  @nrwl/jest : 12.6.3
  @nrwl/linter : 12.6.3
  @nrwl/nest : Not Found
  @nrwl/next : 12.6.3
  @nrwl/node : Not Found
  @nrwl/nx-cloud : Not Found
  @nrwl/react : 12.6.3
  @nrwl/schematics : Not Found
  @nrwl/tao : 12.6.3
  @nrwl/web : 12.6.3
  @nrwl/workspace : 12.6.3
  @nrwl/storybook : 12.6.3
  @nrwl/gatsby : Not Found
  typescript : 4.3.5
  ```

### LINKED ISSUE #6726: Support Next.js 11.1.0
After upgrading to Next.js 11.1.0, executing `nx run website:serve` will lead to `Cannot find module 'next/dist/next-server/lib/constants'`.

### PR DESCRIPTION
**Pull request by rayriffy**
## Current Behavior

Next 11.1.0 has been updated with one breaking change to Nx which is import location of `next-server` (https://github.com/vercel/next.js/pull/26756), which make builds to fail

CleanShot 2021-08-13 at 02 05 10

## Expected Behavior

Build should be able to proceed

- Most of `next/dist/next-server/server` imports can be changed to `next/dist/server`
- `next/dist/next-server/lib/constants` has been moved to `next/dist/shared/lib/constants`

## Related Issue(s)

Fixes #6576

### CODE REVIEW

**Comment by mheob** [comment]
Resolves #6576

**Comment by kirjai** [comment]
thanks a lot for this contribution, really appreciated!

a few points of feedback:

---

a few more places that you need to update the version number in:
- `packages/next/src/utils`
- `packages/next/package.json`

---

with the above changes done - there's actually no need to change the version number in the root `package.json`. that will be updated whenever we update the repo to the latest version of Nx.

---

i see that you've written a migration, which is great! but if all you want to do is update the package version in package.json, then you can do that in `packages/next/migrations.json`, under the `packageJsonUpdates` key:

something similar to
```json
    "12.7.2": {
      "version": "12.7.2-beta.0",
      "packages": {
        "next": {
          "version": "11.1.0",
          "alwaysAddToPackageJson": false
        },
        "eslint-config-next": {
          "version": "11.1.0",
          "alwaysAddToPackageJson": false
        }
      }
    }
```

definitely give me a shout, if you aren't able to implement these changes some time soon, and i'll jump on your branch and add these changes. once all of these are done - we'll see if CI still fails

File: packages/next/src/utils/versions.ts [hunk 1]
```diff
@@ -1,6 +1,6 @@
 export const nxVersion = '*';
 
-export const nextVersion = '11.0.1';
+export const nextVersion = '11.1.0';
 export const eslintConfigNextVersion = '11.0.1';
```
Comment (kirjai) [inline packages/next/src/utils/versions.ts:4]:
```suggestion
export const eslintConfigNextVersion = '11.1.0';
```

Comment (rayriffy) [reply packages/next/src/utils/versions.ts]: Ohhhhh forgot about that one

**Comment by kirjai** [comment]
Apologies, i was wrong, we definitely want to update the version number in the root `package.json` 🤦‍♂️ just pushed a commit with that change to your branch

**Comment by kirjai** [comment]
so we're in an interesting predicament. the version of `next` defined in `package.json` is used both for latest `@nrwl/next` development (preview / what's in the repo right now), as well as by our `nx-dev` project in the repo. additionally, `nx-dev` isn't on the preview version of Nx.. so because of that and because `next` 11.1.0 has a breaking change - bumping the version of `next` that's in the root `package.json` breaks the existing `nx-dev` project.
it's kind of hard to explain in text.. 😄 

but ultimately, i've made some changes so that we can support both versions of `next` at the same time - 11.1.0 and pre 11.1.0..

File: packages/next/src/utils/config.ts [hunk 2]
```diff
@@ -19,10 +21,11 @@ import { normalizeAssets } from '@nrwl/web/src/utils/normalize';
 import { createCopyPlugin } from '@nrwl/web/src/utils/config';
 import { WithNxOptions } from '../../plugins/with-nx';
 import {
-  computeCompilerOptionsPaths,
   createTmpTsConfig,
   DependentBuildableProjectNode,
 } from '@nrwl/workspace/src/utilities/buildable-libs-utils';
+import { importConfig } from './require-shim';
+const { default: loadConfig } = importConfig();
```
Comment (jaysoo) [inline packages/next/src/utils/config.ts:28]: Can just make this 

```
const loadConfig = importConfig();
```

and then in the shim file do

```

export function importConfig() {
  try {
    return require('next/dist/next-server/server/config').default;
  } catch {
    return require('next/dist/server/config').default;
  }
}
```

That way is is more consistent with how you'd  use default imports.

File: packages/next/src/utils/require-shim.ts [hunk 3]
```diff
@@ -0,0 +1,23 @@
...
+
+export function importConfig() {
+  try {
+    return require('next/dist/next-server/server/config');
+  } catch {
+    return require('next/dist/server/config');
+  }
+}
+
+export function importConfigShared() {
```
Comment (jaysoo) [inline packages/next/src/utils/require-shim.ts:17]: I don't think this is used anywhere? Remove?

**Comment by kirjai** [comment]
@rayriffy once again thanks for initiating and helping with this!

**Comment by Joroze** [comment]
> @rayriffy once again thanks for initiating and helping with this!

Thanks Rayriffy! 

I just started using NX. How does the process work for updating deps like NextJS? Do we always have to wait until NX itself supports that new NextJS version itself? Lastly, how do we update NX to use NextJS 11.1?

**Comment by kirjai** [comment]
@Joroze good question. normally, for things like NextJS, since it's a peer dependency - you are free to update it as you need to. but if there are breaking changes between NextJS versions (like in this case, with 11.1) - then we, unfortunately, have to do some work on the Nx side to make it compatible. 

as for updating Nx to use NextJS 11.1, we're going to have a release of Nx soon (12.7.2, i believe). once it's out - you're going to run `nx migrate latest`, and that should do everything, including updating the version of NextJS, in this instance. 
[Documentation on `nx migrate`](https://nx.dev/latest/react/core-concepts/updating-nx).

**Comment by Joroze** [comment]
Thank you @kirjai! This is exactly the answer I needed.

**Comment by kirjai** [comment]
FYI, Nx version `12.8.0-beta.0` is available now that includes Next 11.1.0 support

**Pull request merged**